File: hw/ccedeno/u3/HW29/TemplateMethod/src/main/java/ec/edu/espe/templatemethod/model/CaffeineBeverage.java

package ec.edu.espe.templatemethod.model;

/**
 *
 * @author Andres Cedeno,Code Creators,DCCO-ESPE
 */
public abstract class CaffeineBeverage {
    
    
    public final void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (customerWantsCondiments()) { 
            addCondiments();
        }
    }

    
    void boilWater() {
        System.out.println("Boiling water");
    }

    void pourInCup() {
        System.out.println("Pouring into cup");
    }

    
    protected abstract void brew();
    protected abstract void addCondiments();

    
    protected boolean customerWantsCondiments() {
        return true; // Las subclases pueden sobrescribir este método si no quieren condimentos
    }
}

File: hw/Vaca/u3/HW29TemplateAndCompsite/HW29TemplateMethodJavaScript/model/CaffeineBeverage.js
class CaffeineBeverage {
    prepareRecipe() {
        this.boilWater();
        this.brew();
        this.pourInCup();
        if (this.customerWantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    pourInCup() {
        console.log("Pouring into cup");
    }

    brew() {
        throw new Error("This method must be overridden!");
    }

    addCondiments() {
        throw new Error("This method must be overridden!");
    }

    customerWantsCondiments() {
        return true; 
    }
}
